kayenta-stackdriver/src/test/java/com/netflix/kayenta/stackdriver/metrics/StackdriverMetricsServiceTest.java

```diff
@@ -0,0 +1,153 @@
+package com.netflix.kayenta.stackdriver.metrics;
+
+import static org.assertj.core.api.Assertions.*;
+import static org.mockito.Answers.RETURNS_DEEP_STUBS;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.when;
+
+import com.google.api.services.monitoring.v3.Monitoring;
+import com.google.api.services.monitoring.v3.model.ListMetricDescriptorsResponse;
+import com.google.api.services.monitoring.v3.model.ListTimeSeriesResponse;
+import com.google.api.services.monitoring.v3.model.MetricDescriptor;
+import com.google.api.services.monitoring.v3.model.Point;
+import com.google.api.services.monitoring.v3.model.TimeInterval;
+import com.google.api.services.monitoring.v3.model.TimeSeries;
+import com.google.api.services.monitoring.v3.model.TypedValue;
+import com.netflix.kayenta.canary.CanaryConfig;
+import com.netflix.kayenta.canary.CanaryMetricConfig;
+import com.netflix.kayenta.canary.CanaryScope;
+import com.netflix.kayenta.canary.providers.metrics.StackdriverCanaryMetricSetQueryConfig;
+import com.netflix.kayenta.google.security.GoogleNamedAccountCredentials;
+import com.netflix.kayenta.metrics.MetricSet;
+import com.netflix.kayenta.security.AccountCredentials;
+import com.netflix.kayenta.security.AccountCredentialsRepository;
+import com.netflix.kayenta.stackdriver.canary.StackdriverCanaryScope;
+import com.netflix.spectator.api.Registry;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public class StackdriverMetricsServiceTest {
+
+  private static final String ACCOUNT = "test-account";
+
+  @Mock AccountCredentialsRepository accountCredentialsRepoMock;
+
+  @Mock(answer = RETURNS_DEEP_STUBS)
+  GoogleNamedAccountCredentials googleAccountCredentialsMock;
+
+  @Mock GoogleNamedAccountCredentials stackdriverCredentialsMock;
+  @Mock ListMetricDescriptorsResponse listMetricDescriptorsResponseMock;
+  @Mock StackdriverCanaryScope canaryScopeMock;
+
+  @Mock(answer = RETURNS_DEEP_STUBS)
+  Monitoring monitoringMock;
+
+  @Mock Monitoring.Projects.TimeSeries.List timeSeriesListMock;
+  @Mock ListTimeSeriesResponse responseMock;
+
+  @Mock(answer = RETURNS_DEEP_STUBS)
+  Registry registry;
+
+  @InjectMocks StackdriverMetricsService stackdriverMetricsService;
```

Comment:
This is likely wading too far into personal preference, but I find this way of constructing an object and injecting mocks a bit heavy...or at least too "magic" for me.  I took a crack at a more explicit approach, and found that some of the `@Mock`-annotated objects here are only used in one test, so I went ahead and moved them there.  If you're OK with it, I'll see if I have permissions to push this diff.  If it turns out you really like the `@Mock` + `@InjectMocks` way of doing it, I'm OK to leave it.
```
diff --git a/kayenta-stackdriver/kayenta-stackdriver.gradle b/kayenta-stackdriver/kayenta-stackdriver.gradle
index 3667da5b..3ca3f386 100644
--- a/kayenta-stackdriver/kayenta-stackdriver.gradle
+++ b/kayenta-stackdriver/kayenta-stackdriver.gradle
@@ -3,5 +3,5 @@ dependencies {
   implementation project(":kayenta-google")
   api "com.google.apis:google-api-services-monitoring"
 
-  testImplementation 'org.mockito:mockito-inline:2.13.0'
+  testImplementation 'org.mockito:mockito-inline'
 }
diff --git a/kayenta-stackdriver/src/test/java/com/netflix/kayenta/stackdriver/metrics/StackdriverMetricsServiceTest.java b/kayenta-stackdriver/src/test/java/com/netflix/kayenta/stackdriver/metrics/StackdriverMetricsServiceTest.java
index 2cbaacad..f0afa75b 100644
--- a/kayenta-stackdriver/src/test/java/com/netflix/kayenta/stackdriver/metrics/StackdriverMetricsServiceTest.java
+++ b/kayenta-stackdriver/src/test/java/com/netflix/kayenta/stackdriver/metrics/StackdriverMetricsServiceTest.java
@@ -1,8 +1,8 @@
 package com.netflix.kayenta.stackdriver.metrics;
 
-import static org.assertj.core.api.Assertions.*;
-import static org.mockito.Answers.RETURNS_DEEP_STUBS;
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
 import com.google.api.services.monitoring.v3.Monitoring;
@@ -15,14 +15,13 @@ import com.google.api.services.monitoring.v3.model.TimeSeries;
 import com.google.api.services.monitoring.v3.model.TypedValue;
 import com.netflix.kayenta.canary.CanaryConfig;
 import com.netflix.kayenta.canary.CanaryMetricConfig;
-import com.netflix.kayenta.canary.CanaryScope;
 import com.netflix.kayenta.canary.providers.metrics.StackdriverCanaryMetricSetQueryConfig;
 import com.netflix.kayenta.google.security.GoogleNamedAccountCredentials;
 import com.netflix.kayenta.metrics.MetricSet;
 import com.netflix.kayenta.security.AccountCredentials;
 import com.netflix.kayenta.security.AccountCredentialsRepository;
 import com.netflix.kayenta.stackdriver.canary.StackdriverCanaryScope;
-import com.netflix.spectator.api.Registry;
+import com.netflix.spectator.api.DefaultRegistry;
 import java.io.IOException;
 import java.time.Instant;
 import java.util.ArrayList;
@@ -30,43 +29,37 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.extension.ExtendWith;
-import org.mockito.InjectMocks;
-import org.mockito.Mock;
-import org.mockito.junit.jupiter.MockitoExtension;
+import org.mockito.Mockito;
 
-@ExtendWith(MockitoExtension.class)
 public class StackdriverMetricsServiceTest {
 
   private static final String ACCOUNT = "test-account";
 
-  @Mock AccountCredentialsRepository accountCredentialsRepoMock;
+  private AccountCredentialsRepository accountCredentialsRepoMock;
 
-  @Mock(answer = RETURNS_DEEP_STUBS)
-  GoogleNamedAccountCredentials googleAccountCredentialsMock;
+  private StackdriverMetricsService stackdriverMetricsService;
 
-  @Mock GoogleNamedAccountCredentials stackdriverCredentialsMock;
-  @Mock ListMetricDescriptorsResponse listMetricDescriptorsResponseMock;
-  @Mock StackdriverCanaryScope canaryScopeMock;
-
-  @Mock(answer = RETURNS_DEEP_STUBS)
-  Monitoring monitoringMock;
-
-  @Mock Monitoring.Projects.TimeSeries.List timeSeriesListMock;
-  @Mock ListTimeSeriesResponse responseMock;
-
-  @Mock(answer = RETURNS_DEEP_STUBS)
-  Registry registry;
-
-  @InjectMocks StackdriverMetricsService stackdriverMetricsService;
+  @BeforeEach
+  void setup() {
+    StackdriverMetricsService.StackdriverMetricsServiceBuilder stackdriverMetricsServiceBuilder =
+        StackdriverMetricsService.builder();
+    accountCredentialsRepoMock = mock(AccountCredentialsRepository.class);
+    stackdriverMetricsServiceBuilder.accountCredentialsRepository(accountCredentialsRepoMock).registry(new DefaultRegistry());
+    stackdriverMetricsService = stackdriverMetricsServiceBuilder.build();
+  }
 
   @Test
-  public void readsInt64Metrics() throws IOException {
+  void readsInt64Metrics() throws IOException {
+    GoogleNamedAccountCredentials stackdriverCredentialsMock = mock(GoogleNamedAccountCredentials.class);
     when(accountCredentialsRepoMock.getRequiredOne(ACCOUNT)).thenReturn(stackdriverCredentialsMock);
 
+    Monitoring monitoringMock = mock(Monitoring.class, Mockito.RETURNS_DEEP_STUBS);
     when(stackdriverCredentialsMock.getMonitoring()).thenReturn(monitoringMock);
 
+    Monitoring.Projects.TimeSeries.List timeSeriesListMock = mock(Monitoring.Projects.TimeSeries.List.class);
+
     when(monitoringMock
             .projects()
             .timeSeries()
@@ -79,6 +72,7 @@ public class StackdriverMetricsServiceTest {
             .setIntervalEndTime(anyString()))
         .thenReturn(timeSeriesListMock);
 
+    ListTimeSeriesResponse responseMock = mock(ListTimeSeriesResponse.class);
     when(timeSeriesListMock.execute()).thenReturn(responseMock);
 
     List<TimeSeries> timeSeriesListWithInt64Points = new ArrayList<TimeSeries>();
@@ -109,11 +103,10 @@ public class StackdriverMetricsServiceTest {
                     .build())
             .build();
 
-    CanaryScope canaryScope =
-        new StackdriverCanaryScope()
-            .setStart(Instant.EPOCH)
-            .setEnd(Instant.EPOCH.plusSeconds(1))
-            .setStep(1l);
+    StackdriverCanaryScope canaryScope =
+        new StackdriverCanaryScope();
+    canaryScope.setStart(Instant.EPOCH).setEnd(Instant.EPOCH.plusSeconds(1)).setStep(1l);
+    canaryScope.setProject("my-project");
     List<MetricSet> queriedMetrics =
         stackdriverMetricsService.queryMetrics(
             ACCOUNT, canaryConfig, canaryMetricConfig, canaryScope);
@@ -122,13 +115,16 @@ public class StackdriverMetricsServiceTest {
   }
 
   @Test
-  public void returnsSingleMetricDescriptorInCache() throws IOException {
+  void returnsSingleMetricDescriptorInCache() throws IOException {
+    GoogleNamedAccountCredentials googleAccountCredentialsMock = mock(GoogleNamedAccountCredentials.class, Mockito.RETURNS_DEEP_STUBS);
+
     Set<AccountCredentials> accountCredentialsSetMock = new HashSet<>();
     accountCredentialsSetMock.add(googleAccountCredentialsMock);
 
     when(accountCredentialsRepoMock.getAllOf(AccountCredentials.Type.METRICS_STORE))
         .thenReturn(accountCredentialsSetMock);
 
+    ListMetricDescriptorsResponse listMetricDescriptorsResponseMock = mock(ListMetricDescriptorsResponse.class);
     when(googleAccountCredentialsMock
             .getMonitoring()
             .projects()
```

Reply:
I applied the patch locally it took a closer look at it, it looks great and I like it :)
It's definitely a lot less magic than the decorators and it's much clearer how the mocks are injected this way

Feel free to apply this patch, or do you want me to push a new commit with it?

Reply:
Please go ahead and push a new commit.